### ISSUE TITLE
When creating a bridge interface, the bridge doesn't appear

### ISSUE THREAD

**Issue by mrobbetts** [NixOS/nixpkgs#16230]
## Issue description

I am configuring a new nixOS system. My machine has two ethernet interfaces (`enp3s0` and `enp4s0`) and one wireless interface (`wlp5s0`). I want to bridge the wireless with one of the ethernets, so in my configuration.nix I add:

```
networking = {
    ...
    bridges.br0.interfaces = [ "enp4s0" "wlp5s0" ];
    ...
```

and run `nixos-rebuild switch`.

I observe in the output that, among other things, a
`/nix/store/z18i1q0r4i2bn0vkpcz45d10nn0628nf-unit-br0-netdev.service/br0-netdev.service`
is created. However, `ifconfig` does not show the new bridge interface.

If I manually run:
`systemctl start br0-netdev.service`

the bridge interface appears.

Probably I am doing something wrong, but I would have expected the bridge to appear immediately as a result of `nixos-rebuild switch`.

Can anyone help?

**Comment by danbst** [comment]
I have similar bug with firewall. https://github.com/NixOS/nixpkgs/pull/15815

**Comment by mrobbetts** [comment]
Also, FWIW, if I restart the machine, the bridge is present as intended.

So something seems to not be firing as a result of the `nixos-rebuild switch`

**Comment by danbst** [comment]
Though the fix can be the same - you have to patch nixpkgs file https://github.com/NixOS/nixpkgs/tree/release-16.03/nixos/modules/tasks/network-interfaces.nix, add multi-user.target alongside with network.target

**Comment by mrobbetts** [comment]
@danbst, is that the real fix though? Since multi-user.target depends on network.target itself, isn't this expected to work without adding the extra relationship? It seems redundant I think, doesn't it?

**Comment by danbst** [comment]
Here is my test:

1. add new bridge `br0` to configuration (I've tested with containers' virtual interfaces)
2. `nixos-rebuild test`. Observe nothing in logs `journalctl -u br0-netdev`
3. add this line to configuration: `systemd.services.br0-netdev.wantedBy = [ "multi-user.target" ];`
4. `nixos-rebuild test`. Observe 
```
Jun 19 06:22:56 iron br0-netdev-start[12044]: Adding bridge br0...
Jun 19 06:22:56 iron systemd[1]: Started Bridge Interface br0.
```

> Since multi-user.target depends on network.target itself
I think there are cases when one wants to reach `network.target` but doesn't want to reach `multi-user.target`. That's why there are two targets. Otherwise, yes, `multi-user` includes `network` and `network.target` is redundant.

Though now I don't know whether this is a proper fix. Maybe there is some systemd bug, or maybe there was a systemd bug that is now resolved, I don't know.

I'm on 16.03

**Comment by Nadrieril** [comment]
I have the same problem on current unstable. Apparently nixos-rebuild detects that something has changed in the interfaces because it restarts dhcpcd.service, but it does not start br0-netdev.service...

**Comment by Mic92** [comment]
Disclaimer: I do not like the way proposed by Nixos for network configuration. These brittle generated scripts can not be the replacement for a proper network configuration. 

What I use: Fortunately there is systemd-networkd which provides an answer to almost all configuration issues one might have. In Nixos 16.09 there is already systemd-networkd module, which I however do recommend in the current form for the reasons provided here: https://github.com/NixOS/nixpkgs/pull/20482

In the linked pull request I added an `extraConfig` option which is a much more straight forward way to configure networkd compared to figuring out how to please the incomplete, sometimes wrong option set provided by the module itself. (the pull-request will be available in 17.03 - maybe I could backport it) 

All available options of systemd-networkd can be looked up in the corresponding excellent manpages: `systemd.network` and `system.netdev` including examples.

Bridges can be configured the following way:

```
systemd.network = {
  enable = true;
  netdevs.br0.netdevConfig = {
    Name = "br0";
    Type = "bridge";
  };
  networks = {
    br0.extraConfig = ''
      [Match]
      Name = br0

      [Network]
      DHCP = both
    '';
   enp4s0.extraConfig = ''
     [Match]
     Name=enp4s0

     [Network]
     Bridge=br0
   '';
   wlp5s0.extraConfig = ''
     [Match]
     Name=wlp5s0

     [Network]
     Bridge=br0
   '';
  }
}
```

**Comment by CMCDragonkai** [comment]
On 17.03, using:

```
networking.bridges = {
    br0 = { interfaces = [ "wlp6s0" ]; rtsp = true;
};
networking.interfaces = { 
    br0 = { ipAddress = "10.0.0.1"; prefixLength = 24; ipv6Address = "fd99:cbc4:692::1"; ipv6PrefixLength = 64; };
};
```

Results in some weird behaviour. First running `systemctl status network-addresses-br0.service` says "Dependency failed for Address configuration of br0." and "network-addresses-br0.service: Job network-addresses-br0.service/start failed with result 'dependency'.`. This is probably because br0 is currently down and hasn't been brought up. Then I check `systemctl status br0-netdev.service`, which shows it removing the old bridge, starting the bridge, adding bridge, and then `RTNETLINK answers: Operation not supported`. However I can then manually run `systemctl restart br0-netdev.service` which successfully brings up the bridge, and then run `systemctl restart network-addresses-br0.service` which successfully adds the ipv4 and ipv6 address to the bridge. This problem seems to stem from the `RTNETLINK answers: Operation not supported` which fails to bring up a bridge. However unlike the OP, I did see a Br0 interface. When I reboot, the changes are not persisted, and I have to do the above forced `systemctl` invocations.

I know that I'm bridging only 1 interface here, but even then it should work.

The other thing is that my `wlp6s0` is acquiring an ipv6 address which propagates to br0, which ends up making br0 have 2 ipv6 addresses, not sure if this is related problem or a separate problem. I'd prefer that the bridge only has 2 IP addresses, the ones I set them.
